File: Greedy/JumpGame/Jump_Game.java
// Greedy approach: track furthest reachable index.
// Time: O(n), Space: O(1)
class Solution {
    public boolean canJump(int[] nums) {
        int maxReach = 0;
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            if (i > maxReach) return false;
            maxReach = Math.max(maxReach, i + nums[i]);
            if (maxReach >= n - 1) return true;
        }
        return true;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.canJump(new int[]{2, 3, 1, 1, 4})); // true
        System.out.println(solution.canJump(new int[]{3, 2, 1, 0, 4})); // false
    }
}

File: Greedy/JumpGame/Jump_Game.py
from typing import List

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        """
        Greedy solution: track the furthest reachable index.
        Time: O(n), Space: O(1)
        """
        max_reach = 0
        n = len(nums)
        for i, jump in enumerate(nums):
            if i > max_reach:
                return False
            max_reach = max(max_reach, i + jump)
            if max_reach >= n - 1:
                return True
        return True

if __name__ == "__main__":
    solution = Solution()
    print(solution.canJump([2, 3, 1, 1, 4]))  # True
    print(solution.canJump([3, 2, 1, 0, 4]))  # False
